src/components/ble/weather/WeatherService.cpp

```diff
@@ -42,315 +305,364 @@ namespace Pinetime {
             if (!AddEventToTimeline(std::move(special))) {
               CleanUpQcbor(&decodeContext);
               return BLE_ATT_ERR_INVALID_ATTR_VALUE_LEN;
             }
             break;
           }
           case WeatherData::eventtype::Pressure: {
-            std::unique_ptr<WeatherData::Pressure> pressure = std::make_unique<WeatherData::Pressure>();
-            pressure->timestamp = tmpTimestamp;
-            pressure->eventType = static_cast<WeatherData::eventtype>(tmpEventType);
-            pressure->expires = tmpExpires;
-
-            int64_t tmpPressure = 0;
-            QCBORDecode_GetInt64InMapSZ(&decodeContext, "Pressure", &tmpPressure);
-            if (tmpPressure < 0 || tmpPressure >= 65535) {
+            auto optPressure = Get<uint16_t>(&decodeContext, "Pressure");
+            if (!optPressure) {
               CleanUpQcbor(&decodeContext);
               return BLE_ATT_ERR_INVALID_ATTR_VALUE_LEN;
             }
-            pressure->pressure = tmpPressure; // NOLINT(bugprone-narrowing-conversions,cppcoreguidelines-narrowing-conversions)
 
+            if (useExisting) {
+              CreatePressureEvent(reinterpret_cast<std::unique_ptr<WeatherData::Pressure>&>(existingEvent),
+                                  *optTimestamp,
+                                  *optExpires,
+                                  *optID,
+                                  *optPressure);
+              break;
+            }
+
+            auto pressure = std::make_unique<WeatherData::Pressure>();
+            CreatePressureEvent(pressure, *optTimestamp, *optExpires, *optID, *optPressure);
             if (!AddEventToTimeline(std::move(pressure))) {
               CleanUpQcbor(&decodeContext);
               return BLE_ATT_ERR_INVALID_ATTR_VALUE_LEN;
             }
             break;
           }
           case WeatherData::eventtype::Location: {
-            std::unique_ptr<WeatherData::Location> location = std::make_unique<WeatherData::Location>();
-            location->timestamp = tmpTimestamp;
-            location->eventType = static_cast<WeatherData::eventtype>(tmpEventType);
-            location->expires = tmpExpires;
-
-            UsefulBufC stringBuf; // TODO: Everything ok with lifecycle here?
-            QCBORDecode_GetTextStringInMapSZ(&decodeContext, "Location", &stringBuf);
-            if (UsefulBuf_IsNULLOrEmptyC(stringBuf) != 0) {
+            auto optLocation = Get<std::string>(&decodeContext, "Location");
+            if (!optLocation) {
               CleanUpQcbor(&decodeContext);
               return BLE_ATT_ERR_INVALID_ATTR_VALUE_LEN;
             }
-            location->location = std::string(static_cast<const char*>(stringBuf.ptr), stringBuf.len);
 
-            int64_t tmpAltitude = 0;
-            QCBORDecode_GetInt64InMapSZ(&decodeContext, "Altitude", &tmpAltitude);
-            if (tmpAltitude < -32768 || tmpAltitude >= 32767) {
+            auto optAltitude = Get<int16_t>(&decodeContext, "Altitude");
```

Comment:
Fun fact, this data could be used in the TImer app to adjust the time it takes to boil water for eggs for example.